contracts/vortex/CarbonVortex.sol

```diff
@@ -138,111 +345,505 @@ contract CarbonVortex is ICarbonVortex, Upgradeable, ReentrancyGuardUpgradeable,
+        uint128 targetAmount
+    ) external payable nonReentrant validToken(token) greaterThanZero(targetAmount) {
+        uint128 sourceAmount;
+        if (token == _targetToken) {
+            sourceAmount = _sellFinalTargetForTarget(targetAmount);
+        } else {
+            sourceAmount = _sellTargetTokenForToken(token, targetAmount);
+        }
+        emit TokenTraded({ caller: msg.sender, token: token, sourceAmount: sourceAmount, targetAmount: targetAmount });
+    }
+
+    function _sellTargetTokenForToken(Token token, uint128 targetAmount) private returns (uint128) {
+        uint128 sourceAmount = expectedTradeInput(token, targetAmount);
+        // revert if trade requires 0 target token
+        if (sourceAmount == 0) {
+            revert InvalidTrade();
+        }
+        // check enough target token (if target token is native) has been sent for the trade
+        if (_targetToken == NATIVE_TOKEN && msg.value < sourceAmount) {
+            revert InsufficientNativeTokenSent();
+        }
+        _targetToken.safeTransferFrom(msg.sender, address(this), sourceAmount);
+        // transfer the tokens to caller
+        token.safeTransfer(msg.sender, targetAmount);
+
+        // if the target token is native, refund any excess native token to caller
+        if (_targetToken == NATIVE_TOKEN && msg.value > sourceAmount) {
+            payable(msg.sender).sendValue(msg.value - sourceAmount);
+        }
+
+        // if no final target token is defined, transfer the target token to `transferAddress`
+        if (Token.unwrap(_finalTargetToken) == address(0)) {
+            // safe due to nonreenrant modifier (forwards all available gas if token is native)
+            _targetToken.unsafeTransfer(_transferAddress, sourceAmount);
+            // increment total collected in `transferAddress`
+            _totalCollected += sourceAmount;
+        }
+
+        // if remaining balance is below the min token sale amount, reset the auction
+        if (_amountAvailableForTrading(token) < _minTokenSaleAmounts[token]) {
+            _resetTrading(token, 0);
+        }
+
+        return sourceAmount;
+    }
+
+    function _sellFinalTargetForTarget(uint128 targetAmount) private returns (uint128) {
+        uint128 sourceAmount = expectedTradeInput(_targetToken, targetAmount);
+        // revert if trade requires 0 finalTarget tokens
+        if (sourceAmount == 0) {
+            revert InvalidTrade();
+        }
+
+        // check enough final target token (if final target token is native) has been sent for the trade
+        if (_finalTargetToken == NATIVE_TOKEN) {
+            if (msg.value < sourceAmount) {
+                revert InsufficientNativeTokenSent();
+            }
+            payable(_transferAddress).sendValue(sourceAmount);
+        } else {
+            // transfer the tokens from the user to the _transferAddress
+            _finalTargetToken.safeTransferFrom(msg.sender, _transferAddress, sourceAmount);
```

Comment:
`_finalTargetToken` can be `0x0` and then `_sellFinalTargetForTarget` will revert, in which case `trade` will never work

Reply:
if `finalTargetToken` is `0x0`, we don't want `sellFinalTargetForTarget` to work - we directly transfer the `targetToken` to the `transferAddress` when it's:
1) withdrawn using `execute`
2) traded via `_sellTargetTokenForToken`

Reply:
So it's ok that `trade` won't work in this case either?

Reply:
Yes